module/zfs/dmu_send.c

```diff
@@ -934,7 +967,7 @@ dmu_send_impl(void *tag, dsl_pool_t *dp, dsl_dataset_t *to_ds,
     zfs_bookmark_phys_t *ancestor_zb, boolean_t is_clone,
     boolean_t embedok, boolean_t large_block_ok, boolean_t compressok,
     boolean_t rawok, int outfd, uint64_t resumeobj, uint64_t resumeoff,
-    vnode_t *vp, offset_t *off)
+    vnode_t *vp, offset_t *off, uint64_t incremental_maxobj)
```

Comment:
I think this function can figure out `incremental_maxobj` instead of making the caller pass it in.  In addition to making the interface cleaner, it looks like we can eliminate some duplicated code in the callers.

Reply:
but we need the maxobj of the fromsnap dataset (because those are the objects we potentially free, unless I misunderstood something..), and here we only have the ancestor_zb (which could be a bookmark!). I figured we would not want to duplicate all the dataset finding code from dmu_send(_obj) in dmu_send_impl ? but maybe I am just missing the right helper to get from zfs_bookmark_phys_t to dnode_phys_t ;)

Reply:
or should we use tosnap to get maxblkid, because it is guarantueed that maxblkid does not get smaller from one snapshot to the next (not sure about this?) - that would indeed make it easier, and also be reusable for the full send proposal..